src/main/java/ch/njol/skript/expressions/ExprTarget.java

```diff
@@ -63,84 +68,93 @@ public class ExprTarget extends PropertyExpression<LivingEntity, Entity> {
 	}
 
 	@Nullable
-	EntityData<?> type;
+	private EntityData<?> type;
 
-	@SuppressWarnings("unchecked")
 	@Override
+	@SuppressWarnings("unchecked")
 	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parser) {
 		type = exprs[matchedPattern] == null ? null : (EntityData<?>) exprs[matchedPattern].getSingle(null);
 		setExpr((Expression<? extends LivingEntity>) exprs[1 - matchedPattern]);
 		return true;
 	}
 
 	@Override
-	protected Entity[] get(Event e, LivingEntity[] source) {
+	protected Entity[] get(Event event, LivingEntity[] source) {
 		return get(source, new Converter<LivingEntity, Entity>() {
 			@Override
 			@Nullable
-			public Entity convert(LivingEntity en) {
-				if (getTime() >= 0 && e instanceof EntityTargetEvent && en.equals(((EntityTargetEvent) e).getEntity()) && !Delay.isDelayed(e)) {
-					Entity target = ((EntityTargetEvent) e).getTarget();
+			public Entity convert(LivingEntity entity) {
+				if (event instanceof EntityTargetEvent && entity.equals(((EntityTargetEvent) event).getEntity()) && !Delay.isDelayed(event)) {
+					Entity target = ((EntityTargetEvent) event).getTarget();
 					if (target == null || type != null && !type.isInstance(target))
 						return null;
 					return target;
 				}
-				return getTarget(en, type);
+				return getTarget(entity, type);
 			}
 		});
 	}
 
 	@Override
-	public Class<? extends Entity> getReturnType() {
-		return type != null ? type.getType() : Entity.class;
+	@Nullable
+	public Class<?>[] acceptChange(ChangeMode mode) {
+		switch (mode) {
+			case SET:
+			case RESET:
+			case DELETE:
+				return CollectionUtils.array(LivingEntity.class);
+		}
+		return super.acceptChange(mode);
 	}
 
 	@Override
-	public String toString(@Nullable Event e, boolean debug) {
-		if (e == null)
-			return "the target" + (type == null ? "" : "ed " + type) + (getExpr().isDefault() ? "" : " of " + getExpr().toString(e, debug));
-		return Classes.getDebugMessage(getAll(e));
+	public void change(Event event, @Nullable Object[] delta, ChangeMode mode) {
+		if (mode == ChangeMode.SET || mode == ChangeMode.RESET || mode == ChangeMode.DELETE) { // null will make the entity target-less (reset target) but for players it will remove them
+			LivingEntity target = delta == null ? null : (LivingEntity) delta[0];
+			for (LivingEntity entity : getExpr().getArray(event)) {
+				if (event instanceof EntityTargetEvent && entity.equals(((EntityTargetEvent) event).getEntity()) && !Delay.isDelayed(event)) {
```

Comment:
```diff
+				if (event instanceof EntityTargetEvent && entity.equals(((EntityTargetEvent) event).getEntity())) {
```
Delay checking is not required when the setTime() method is properly implemented.

See my comment below for information.

This suggestion gets suppressed by my other suggestion which re-writes this for loop.